src/touch.rs

```diff
@@ -0,0 +1,102 @@
+use esp_idf_sys::{
+    esp, touch_fsm_mode_t, touch_fsm_mode_t_TOUCH_FSM_MODE_MAX, touch_fsm_mode_t_TOUCH_FSM_MODE_SW,
+    touch_fsm_mode_t_TOUCH_FSM_MODE_TIMER, touch_pad_config, touch_pad_fsm_start, touch_pad_init,
+    touch_pad_read_raw_data, touch_pad_set_fsm_mode, touch_pad_t, touch_pad_t_TOUCH_PAD_NUM0,
+    touch_pad_t_TOUCH_PAD_NUM1, touch_pad_t_TOUCH_PAD_NUM10, touch_pad_t_TOUCH_PAD_NUM11,
+    touch_pad_t_TOUCH_PAD_NUM12, touch_pad_t_TOUCH_PAD_NUM13, touch_pad_t_TOUCH_PAD_NUM14,
+    touch_pad_t_TOUCH_PAD_NUM2, touch_pad_t_TOUCH_PAD_NUM3, touch_pad_t_TOUCH_PAD_NUM4,
+    touch_pad_t_TOUCH_PAD_NUM5, touch_pad_t_TOUCH_PAD_NUM6, touch_pad_t_TOUCH_PAD_NUM7,
+    touch_pad_t_TOUCH_PAD_NUM8, touch_pad_t_TOUCH_PAD_NUM9, EspError,
+};
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub enum FsmMode {
+    Timer,
+    SW,
+    Max,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl From<FsmMode> for touch_fsm_mode_t {
+    fn from(value: FsmMode) -> Self {
+        match value {
+            FsmMode::Timer => touch_fsm_mode_t_TOUCH_FSM_MODE_TIMER,
+            FsmMode::SW => touch_fsm_mode_t_TOUCH_FSM_MODE_SW,
+            FsmMode::Max => touch_fsm_mode_t_TOUCH_FSM_MODE_MAX,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub enum TouchPad {
+    Pad0,
+    Pad1,
+    Pad2,
+    Pad3,
+    Pad4,
+    Pad5,
+    Pad6,
+    Pad7,
+    Pad8,
+    Pad9,
+    Pad10,
+    Pad11,
+    Pad12,
+    Pad13,
+    Pad14,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl From<TouchPad> for touch_pad_t {
+    fn from(value: TouchPad) -> Self {
+        match value {
+            TouchPad::Pad0 => touch_pad_t_TOUCH_PAD_NUM0,
+            TouchPad::Pad1 => touch_pad_t_TOUCH_PAD_NUM1,
+            TouchPad::Pad2 => touch_pad_t_TOUCH_PAD_NUM2,
+            TouchPad::Pad3 => touch_pad_t_TOUCH_PAD_NUM3,
+            TouchPad::Pad4 => touch_pad_t_TOUCH_PAD_NUM4,
+            TouchPad::Pad5 => touch_pad_t_TOUCH_PAD_NUM5,
+            TouchPad::Pad6 => touch_pad_t_TOUCH_PAD_NUM6,
+            TouchPad::Pad7 => touch_pad_t_TOUCH_PAD_NUM7,
+            TouchPad::Pad8 => touch_pad_t_TOUCH_PAD_NUM8,
+            TouchPad::Pad9 => touch_pad_t_TOUCH_PAD_NUM9,
+            TouchPad::Pad10 => touch_pad_t_TOUCH_PAD_NUM10,
+            TouchPad::Pad11 => touch_pad_t_TOUCH_PAD_NUM11,
+            TouchPad::Pad12 => touch_pad_t_TOUCH_PAD_NUM12,
+            TouchPad::Pad13 => touch_pad_t_TOUCH_PAD_NUM13,
+            TouchPad::Pad14 => touch_pad_t_TOUCH_PAD_NUM14,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct TouchDriver {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl TouchDriver {
+    pub fn new() -> Result<Self, EspError> {
+        Ok(TouchDriver {})
+    }
+
+    pub fn init(&mut self) -> Result<(), EspError> {
```

Comment:
in some drivers there is a compelling reason to split init from the creation of the driver - but is this the case for the touch driver? Same goes for calling the config step separately. I think it would be nicer to just call TouchDriver::new(&some_config)?; and be done with all initial steps. Reducing the api surface the user needs to interact on and less chance to do things out of order.

Reply:
Yea happy to consolidated new + init + config, should the FSM stuff go in there as well? I'm not sure I quite understand what it does...

Reply:
its a onetime config that is not supposed to be changed after init if so i also would combine it

Reply:
hmmm in the case that the FSM mode is set to software a different start function should be used, not sure I'm understanding it right but seem that the point of this is so you can control when it starts in software, and so maybe it shouldn't be part of `new`? Or have I misunderstood?